browser/ephemeral_storage/ephemeral_storage_browsertest.cc

```diff
@@ -253,41 +262,66 @@ IN_PROC_BROWSER_TEST_F(EphemeralStorageBrowserTest, StorageIsPartitioned) {
-  EXPECT_EQ("a.com value", values_before.iframe_2.local_storage);
+  ValuesFromFrames values = GetValuesFromFrames(web_contents);
+  EXPECT_EQ("a.com value", values.main_frame.local_storage);
+  EXPECT_EQ("a.com value", values.iframe_1.local_storage);
+  EXPECT_EQ("a.com value", values.iframe_2.local_storage);
 
-  EXPECT_EQ("a.com value", values_before.main_frame.session_storage);
-  EXPECT_EQ("a.com value", values_before.iframe_1.session_storage);
-  EXPECT_EQ("a.com value", values_before.iframe_2.session_storage);
+  EXPECT_EQ("a.com value", values.main_frame.session_storage);
+  EXPECT_EQ("a.com value", values.iframe_1.session_storage);
+  EXPECT_EQ("a.com value", values.iframe_2.session_storage);
 
-  EXPECT_EQ("from=a.com", values_before.main_frame.cookies);
-  EXPECT_EQ("from=a.com", values_before.iframe_1.cookies);
-  EXPECT_EQ("from=a.com", values_before.iframe_2.cookies);
+  EXPECT_EQ("from=a.com", values.main_frame.cookies);
+  EXPECT_EQ("from=a.com", values.iframe_1.cookies);
+  EXPECT_EQ("from=a.com", values.iframe_2.cookies);
 
   // Navigate away and then navigate back to the original site.
   ui_test_utils::NavigateToURL(browser(), b_site_ephemeral_storage_url_);
+  ASSERT_TRUE(WaitForLoadStop(web_contents));
   ui_test_utils::NavigateToURL(browser(), a_site_ephemeral_storage_url_);
+  ASSERT_TRUE(WaitForLoadStop(web_contents));
 
-  ValuesFromFrames values_after = GetValuesFromFrames(web_contents);
-  EXPECT_EQ("a.com value", values_after.main_frame.local_storage);
-  EXPECT_EQ(nullptr, values_after.iframe_1.local_storage);
-  EXPECT_EQ(nullptr, values_after.iframe_2.local_storage);
+  // within keepalive values should be the same
+  ValuesFromFrames before_timeout = GetValuesFromFrames(web_contents);
+  EXPECT_EQ("a.com value", before_timeout.main_frame.local_storage);
+  EXPECT_EQ("a.com value", before_timeout.iframe_1.local_storage);
+  EXPECT_EQ("a.com value", before_timeout.iframe_2.local_storage);
 
-  EXPECT_EQ("a.com value", values_after.main_frame.session_storage);
-  EXPECT_EQ(nullptr, values_after.iframe_1.session_storage);
-  EXPECT_EQ(nullptr, values_after.iframe_2.session_storage);
+  // keepalive does not apply to session storage
+  EXPECT_EQ("a.com value", before_timeout.main_frame.session_storage);
+  EXPECT_EQ(nullptr, before_timeout.iframe_1.session_storage);
+  EXPECT_EQ(nullptr, before_timeout.iframe_2.session_storage);
 
-  EXPECT_EQ("from=a.com", values_after.main_frame.cookies);
-  EXPECT_EQ("", values_after.iframe_1.cookies);
-  EXPECT_EQ("", values_after.iframe_2.cookies);
+  EXPECT_EQ("from=a.com", before_timeout.main_frame.cookies);
+  EXPECT_EQ("from=a.com", before_timeout.iframe_1.cookies);
+  EXPECT_EQ("from=a.com", before_timeout.iframe_2.cookies);
+
+  // after keepalive values should be cleared
+  ui_test_utils::NavigateToURL(browser(), b_site_ephemeral_storage_url_);
+  ASSERT_TRUE(WaitForLoadStop(web_contents));
+  base::PlatformThread::Sleep(base::TimeDelta::FromSeconds(kKeepAliveInterval));
```

Comment:
this way you hang the whole process. consider this instead:
```
base::RunLoop run_loop;
base::SequencedTaskRunnerHandle::Get()->PostDelayedTask(
    FROM_HERE, run_loop.QuitClosure(),
    base::TimeDelta::FromSeconds(kKeepAliveInterval));
run_loop.Run();
```

Reply:
good call, I was digging around trying to remember how to wait and saw other tests using `base::PlatformThread::Sleep` so I thought I had misremembered that there was a better way :)

Reply:
but why not `ScopedMockTimeMessageLoopTaskRunner` ?

Reply:
I wasn't aware of its existence, but looks like a good option because it doesn't require waiting at all and also eliminates a potential race condition if the tests are very slow